nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/security/SecurityService.java

```diff
@@ -156,6 +157,40 @@ public boolean checkPermission(Document doc, Principal principal, String permiss
     }
 
     /**
+     * Filters the supplied permissions base on whether they are granted to a given principal for a given document.
+     *
+     * @since 9.1
+     */
+    public String[] filterGrantedPermissions(Document doc, Principal principal, String... permissions) {
+        String username = principal.getName();
+
+        if (SecurityConstants.SYSTEM_USERNAME.equals(username)) {
+            return permissions;
+        }
+
+        if (principal instanceof NuxeoPrincipal && ((NuxeoPrincipal) principal).isAdministrator()) {
+            return permissions;
+        }
+
+        String[] additionalPrincipals = getPrincipalsToCheck(principal);
+        ACP acp = doc.getSession().getMergedACP(doc);
+
+        List<String> result = new ArrayList<>();
+        for(String permission : permissions) {
+            String[] resolvedPermissions = getPermissionsToCheck(permission);
+            Access access = securityPolicyService.checkPermission(doc, acp, principal, permission, resolvedPermissions,
+                additionalPrincipals);
+            if (access == null || Access.UNKNOWN.equals(access)) {
+                access = (acp != null) ? acp.getAccess(additionalPrincipals, resolvedPermissions) : null;
```

Comment:
I prefer to write `acp == null ? null : acp.getAccess(additionalPrincipals, resolvedPermissions)` to avoid negations.